### PR TITLE
Modular Jsoo settings [WIP]

### PR DESCRIPTION
**Pull request by rgrinberg**
WIP to address #970 

I've added the basic functionality that `--dev` enabled via the single `jsoo_compilation` field. However, the old mode is a bit more involved than just separate compilation and also involved some flag switching. The flags thing I think is best handled the same way we handle `ocaml_flags`. So either include the profile dependant jsoo flags in `Ocaml_flags.t` or make something similar.

I'm also taking suggestions for naming. What I think would be ideal if we had a field for env that looked like this:

```
(js_of_ocaml
 (compilation separate|classic)
 (flags <osl>))
```

Btw: `js_of_ocaml` is kind of a mouthful, I'd be in favor of deprecating it and adopting `jsoo` instead in the dune lang.

### CODE REVIEW

File: src/js_of_ocaml_rules.ml [hunk 1]
```diff
@@ -4,14 +4,19 @@ open Build.O
 
 module SC = Super_context
 
-let dev_mode sctx = SC.profile sctx = "dev"
+let jsoo_compilation = SC.Env.jsoo_compilation
 
-let separate_compilation_enabled = dev_mode
+let pretty sctx ~dir =
+  match jsoo_compilation sctx ~dir with
+  | Separate ->  ["--pretty"]
```
Comment (hhugo) [inline src/js_of_ocaml_rules.ml]: `--pretty` and `--sourcemap` have no relationship with separate compilation. There should be enabled with the dev profile.

Comment (rgrinberg) [reply src/js_of_ocaml_rules.ml]: Yup that is true. I just kept the ild behavior to keep things simple for an initial PR. As I said in my ticket, the plan is to treat these as normal default flags for dev profile

**Comment by rgrinberg** [comment]
@diml when you're back, can you add hhugo to the team of maintainers as the jsoo maintainer? I could not select him from the list of reviewers.

@hhugo What do you think about splitting the jsoo flags field into links and compile flags in analogy to the regular compile flags. This way we'll be able to set `--source-map-inline` only to `jsoo_link` and then we'll be able to have proper profile dependant defaults.

**Comment by rgrinberg** [comment]
Depends on #1048

**Comment by hhugo** [comment]
@rgrinberg, `--source-map-inline` is also needed when compiling.

**Comment by rgrinberg** [comment]
Understood, but we still need separate fields for compilation and linking. Otherwise there won't be a way to set the defaults such that `--pretty` isn't passed to jsoo_link.

**Comment by hhugo** [comment]
Well, it's a big more complicated.
- `js_of_ocaml` (the binary) can do both "compilation" and "linking".
- `jsoo_link` is only use to link js files together in separate compilation.
- some `jsoo_link` flags cannot be pass to `js_of_ocaml`

**Comment by diml** [comment]
I added @hhugo

**Comment by rgrinberg** [comment]
Thanks for clarifying @hhugo, this does complicate things a little bit. Essentially, the link flags would be entirely useless separate compilation was turned on. I see possibilities such as:

* Making it clear to the user that link flags are actually jsoo_link flags with a `jsoo_link` flag. But I don't like this, because it's a bit too low level of a detail for users.

* adding a separate compilation flags field such as:

```
(separate_compilation_flags
 (compile <flags>)
 (link <flags))
```

These flags would only be consumed in separate compilation mode of course.

* Sticking to my original proposal but simply ignoring the link flags when separate compilation is off. The only issue here is that a user might not expect the link flags to be ignored but it's not so bad.

Any other ideas you have in mind? Or a preference to any of the above?

**Comment by pbiggar** [comment]
@rgrinberg asked me to comment on this proposal as I have a related request in #1613.

I think it's good. I tried to use this functionality without realizing it didn't exist, and the thing I tried looks a lot like this.

The one comment I have is on the naming of the `compilation` flag. I think `classic` doesn't mean anything. I would rename it. Perhaps an antonym to "separate" such as "together" or "single", or a term from traditional compilation like "whole-program". Hope that's not too bikeshedy.

**Comment by Lupus** [comment]
Has this been considered a go or no go? Setting `--enable=with-js-error --enable=excwrap` globally is still not possible without pinning js_of_ocaml with this option changed in-code...

**Comment by hhugo** [comment]
I'm looking at this now.
Will work on https://github.com/ocaml/dune/pull/5049

**Comment by rgrinberg** [comment]
Superseded by #5049

**Pull request closed without merging**